File: frontend/lib/services/settings_service.dart
import 'package:flutter/foundation.dart';

import '../ffi/protonctl_bridge.dart';
import '../models/app_config.dart';
import '../models/steam_candidate.dart';

/// Wraps the native `AppConfig` (Steam path, compatibilitytools.d
/// location, first-start flag, persisted theme choices) so the rest of
/// the app can read/write it as a normal Dart object instead of poking
/// JSON through the FFI bridge directly.
class SettingsService extends ChangeNotifier {
  AppConfig _config = AppConfig();
  AppConfig get config => _config;

  Future<void> load() async {
    _config = AppConfig.fromJson(ProtonCtlBridge.loadConfig());
    notifyListeners();
  }

  Future<void> update(AppConfig Function(AppConfig current) updater) async {
    _config = updater(_config);
    ProtonCtlBridge.saveConfig(_config.toJson());
    notifyListeners();
  }

  Future<void> completeFirstStart({
    required String steamPath,
    required String compatibilityToolsDir,
    required SteamInstallType installType,
  }) async {
    await update((c) => c.copyWith(
          steamPath: steamPath,
          compatibilityToolsDir: compatibilityToolsDir,
          installType: installType,
          firstStartCompleted: true,
        ));
  }
}

File: frontend-compose/src/main/kotlin/com/protonctl/app/service/SettingsService.kt
package com.protonctl.app.service

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.protonctl.app.model.AppConfig
import com.protonctl.app.model.SteamInstallType
import com.protonctl.app.native.ProtonCtlBridge

/** Wraps the native [AppConfig] so the rest of the app reads/writes it as
 * plain Compose state instead of poking JSON through the bridge directly. */
class SettingsService {
    var config by mutableStateOf(AppConfig())
        private set

    fun load() {
        config = ProtonCtlBridge.loadConfig()
    }

    fun update(updater: (AppConfig) -> AppConfig) {
        config = updater(config)
        ProtonCtlBridge.saveConfig(config)
    }

    fun completeFirstStart(steamPath: String, compatibilityToolsDir: String, installType: SteamInstallType) {
        update {
            it.copy(
                steamPath = steamPath,
                compatibilityToolsDir = compatibilityToolsDir,
                installType = installType,
                firstStartCompleted = true,
            )
        }
    }
}
